src/core/execute.c

```diff
@@ -2627,84 +2791,54 @@ static int acquire_credentials(
-
-                if (path_is_absolute(lc->path)) {
-                        /* If this is an absolute path, read the data directly from it, and support AF_UNIX sockets */
-                        source = lc->path;
-                        flags |= READ_FULL_FILE_CONNECT_SOCKET;
-
-                        /* Pass some minimal info about the unit and the credential name we are looking to acquire
-                         * via the source socket address in case we read off an AF_UNIX socket. */
-                        if (asprintf(&bindname, "@%" PRIx64"/unit/%s/%s", random_u64(), unit, lc->id) < 0)
-                                return -ENOMEM;
-
-                        missing_ok = false;
+                _cleanup_close_ int sub_fd = -1;
 
-                } else if (params->received_credentials) {
-                        /* If this is a relative path, take it relative to the credentials we received
-                         * ourselves. We don't support the AF_UNIX stuff in this mode, since we are operating
-                         * on a credential store, i.e. this is guaranteed to be regular files. */
-                        j = path_join(params->received_credentials, lc->path);
-                        if (!j)
-                                return -ENOMEM;
-
-                        source = j;
-                } else
-                        source = NULL;
-
-                if (source)
-                        r = read_full_file_full(
-                                        AT_FDCWD, source,
-                                        UINT64_MAX,
-                                        lc->encrypted ? CREDENTIAL_ENCRYPTED_SIZE_MAX : CREDENTIAL_SIZE_MAX,
-                                        flags | (lc->encrypted ? READ_FULL_FILE_UNBASE64 : 0),
-                                        bindname,
-                                        &data, &size);
-                else
-                        r = -ENOENT;
-                if (r == -ENOENT && (missing_ok || hashmap_contains(context->set_credentials, lc->id))) {
-                        /* Make a missing inherited credential non-fatal, let's just continue. After all apps
-                         * will get clear errors if we don't pass such a missing credential on as they
-                         * themselves will get ENOENT when trying to read them, which should not be much
-                         * worse than when we handle the error here and make it fatal.
-                         *
-                         * Also, if the source file doesn't exist, but a fallback is set via SetCredentials=
-                         * we are fine, too. */
-                        log_debug_errno(r, "Couldn't read inherited credential '%s', skipping: %m", lc->path);
+                if (STR_IN_SET(lc->path,
+                               "passwd.hashed-password.root",
+                               "passwd.plaintext-password.root",
+                               "passwd.shell.root"))
```

Comment:
where are these coming from?

Reply:
Tests were failing on `systemd-sysusers.service`, if you look in this file you'll find these three credentials which for some reason have no path specified (which is apparently valid syntax).

```
# Optionally, pick up a root password and shell for the root user from a
# credential passed to the service manager. This is useful for importing this
# data from nspawn's --set-credential= switch.
LoadCredential=passwd.hashed-password.root
LoadCredential=passwd.plaintext-password.root
LoadCredential=passwd.shell.root
```

Oddly enough, if I try this in `systemd-run`, it fails:
```
$ systemd-run -p "LoadCredential=passwd.shell.root" echo "OK"
Missing argument to LoadCredential=.
```
Not really sure what to make of this, but in any case, when the path is left unspecified for these credentials, it ends up as the value of the ID, so I skip over these whenever they show up as path. This makes all the tests pass.

Reply:
It's this case defined in the manpage, no?

```
If the specified path is not absolute and itself qualifies as valid credential identifier it
is understood to refer to a credential that the service manager itself received via the
$CREDENTIALS_DIRECTORY environment variable, which may be used to propagate credentials from an invoking
environment (e.g. a container manager that invoked the service manager) into a service. <....> If
the file system path is omitted it is chosen identical to the credential name, i.e. this is a terse way
do declare credentials to inherit from the service manager into a service.
```

It doesn't seem right to hard-code those values, rather, this generic use case should be supported fully. In other words, if someone is using it to pass stuff via nspawn, we can't break it.

Reply:
Makes sense, didn't realize this was a general case